### ISSUE TITLE
Sample from a subset of the token_embedding_table

### ISSUE THREAD

**Issue by PLarsen79** [karpathy/nanoGPT#442]
I wonder what to change in the code if I at inference time only want the logits and sample from the probabilities distribution on a subset of the total vocabulary in the token_embedding_table?

I don't seek a post-processing solution where the probabilities are altered after the logits are obtained, but instead I want a more effective smaller set of tokens to chose from and not the total vocabulary, if that makes any sense.

**Comment by VatsaDev** [comment]
change tokenizer and retrain ig?

**Comment by PLarsen79** [comment]
I am not sure I understand your suggestion, perhaps I should clarify my question a bit.
My goal is to be able to have, for simplicity say two separate vocabularies inside the same model, where the first can be the char-vocab as in vanilla nanoGPT and the second can be a special-token vocab. The data can be setup to always have a special token every 10th token on the prompt, training data and sampling.

An example: "This is a[s] normal s[s2]entence." where [s] and [s2] are two token from a separate vocabulary. They can be [1] and [2].. and in the char-vocab there could also be token [1] and [2], but they are trained individually.

So when sampling, in this example, the model should sample only from the special-token vocab every 10th token. All training data is also constructed in the same way.

I figured out I probably should put a separate vocabulary in the GPT class init, something like this:

```
    self.transformer = nn.ModuleDict(dict(
            wte = nn.Embedding(config.vocab_size, config.n_embd),
            wt2e = nn.Embedding(config.vocab2_size, config.n_embd),
            wpe = nn.Embedding(config.block_size, config.n_embd),
```

But then I kind of fail to break down the forward function, assuming I guess it's too optimized for a single vocab architecture.
My question breaks down to what changes I need to do:

1.  here (last line) in for forward function:

```
    def forward(self, idx, targets=None):
        ....

        # forward the GPT model itself
        tok_emb = self.transformer.wte(idx) # token embeddings of shape (b, t, n_embd)
        pos_emb = self.transformer.wpe(pos) # position embeddings of shape (t, n_embd)
        x = self.transformer.drop(tok_emb + pos_emb)

```

2. and also here when retrieving the logits and probabilities for every 10th token:

```
        if targets is not None:
            # if we are given some desired targets also calculate the loss
            logits = self.lm_head(x)
            loss = F.cross_entropy(logits.view(-1, logits.size(-1)), targets.view(-1), ignore_index=-1)
        else:
            # inference-time mini-optimization: only forward the lm_head on the very last position
            logits = self.lm_head(x[:, [-1], :]) # note: using list [-1] to preserve the time dim
            loss = None
```

I already know how to add special tokens to for example a char-vocab, but that's not what I am looking for. 
I am looking for a way to train and to sample from many different (smaller) vocabularies.

The two vocabulary is just an example to keep my question simple, I am actually trying to use about 100 different vocabs, where each contains about 100 special tokens. Instead of having a single 10k token vocabulary.

**Comment by VatsaDev** [comment]
Oh i thought you wanted to expand the tokenizer, I've never though of or heard of two tokenizers in the model at different points, except for multimodality? Might look into that
